Metalama.Framework.Engine/Linking/LinkerRewritingDriver.Types.cs

```diff
@@ -0,0 +1,179 @@
+        {
+#if ROSLYN_4_8_0_OR_GREATER
+            structDeclaration =
+                structDeclaration.PartialUpdate(
+                    attributeLists: RewritePrimaryConstructorTypeAttributeLists( structDeclaration.AttributeLists ),
+                    parameterList: default( ParameterListSyntax ) );
+#else
+        throw new AssertionFailedException( "This code should not run in this Roslyn version." );
+#endif
+        }
+
+        structDeclaration = structDeclaration.WithMembers( List( transformedMembers ) );
+
+        return structDeclaration;
+    }
+
+    public RecordDeclarationSyntax RewriteRecord(
+        RecordDeclarationSyntax recordDeclaration,
+        INamedTypeSymbol symbol,
+        IReadOnlyList<MemberDeclarationSyntax> transformedMembers )
+    {
+        if ( this.LateTransformationRegistry.HasRemovedPrimaryConstructor( symbol ) )
+        {
+            recordDeclaration =
+                recordDeclaration.PartialUpdate(
+                    attributeLists: RewritePrimaryConstructorTypeAttributeLists( recordDeclaration.AttributeLists ),
+                    parameterList: default( ParameterListSyntax ),
+                    baseList:
+                        recordDeclaration.BaseList != null
+                        ? recordDeclaration.BaseList.WithTypes(
+                            SeparatedList(
+                            recordDeclaration.BaseList.Types.SelectAsArray( b =>
+                                b switch
+                                {
+                                    PrimaryConstructorBaseTypeSyntax pc => SimpleBaseType( pc.Type ),
+                                    _ => b
+                                } ) ) )
+                        : default );
+        }
+        else if ( recordDeclaration.ParameterList != null )
+        {
+            var semanticModel = this.IntermediateCompilationContext.SemanticModelProvider.GetSemanticModel( recordDeclaration.SyntaxTree );
+            SyntaxGenerationContext? generationContext = null;
+
+            List<MemberDeclarationSyntax>? newMembers = null;
+
+            var transformedParametersAndCommas = new List<SyntaxNodeOrToken>( recordDeclaration.ParameterList.Parameters.Count * 2 );
+
+            for ( var i = 0; i < recordDeclaration.ParameterList.Parameters.Count; i++ )
+            {
+                var parameter = recordDeclaration.ParameterList.Parameters[i];
+                newMembers ??= new List<MemberDeclarationSyntax>();
+
+                var parameterSymbol = semanticModel.GetDeclaredSymbol( parameter );
+
+                if ( parameterSymbol == null )
+                {
+                    continue;
+                }
+
+                var propertySymbol = parameterSymbol.ContainingType.GetMembers( parameterSymbol.Name ).OfType<IPropertySymbol>().FirstOrDefault();
+
+                if ( propertySymbol != null && this.IsRewriteTarget( propertySymbol ) )
```

Comment:
How is this going to behave if the record has a property with the same name as a parameter? E.g.:

```c#
record R(int I, int J)
{
	int I { get; set; }
}
```

I think the two cases can be distinguished using `DeclaringSyntaxReferences`.

Reply:
```
record R([property: Attr]int P);
```

and

```
record R(int P)
{
    [Attr]
    public int P {get;init;} = P;
}
```

Should be equivalent. But I see the code is wrong there.

Reply:
Created a follow up bug (this I only moved this code in this PR).